### ISSUE TITLE
RestoreAssets undefined on example 9 (bootstrap)

### ISSUE THREAD

**Issue by prsolucoes** [asticode/go-astilectron#10]
The line 45 in "examples/9.bootstrap/main.go" is:

```
RestoreAssets: RestoreAssets,
```

But need be something like this:

```
RestoreAssets: func(dir, name string) error {
    return nil
},
```

**Comment by asticode** [comment]
Mmm in this example I wanted to show that you could disembed assets as well using a legit `RestoreAssets` signature function. In the `readme.md` I added that you should run the following commands:

    $ go generate examples/9.bootstrap/main.go
    $ go run examples/9.bootstrap/main.go examples/9.bootstrap/resources.go -v

so that it could generate the `resources.go` file thanks to `go-bindata` and so that you could use its `RestoreAssets` function.

For this example I don't want to put an empty `RestoreAssets` as you suggested.

Does it make sense?

**Comment by prsolucoes** [comment]
Yes, but i think you need put this file inside example folder, no?

**Comment by asticode** [comment]
I don't think committing the file `resources.go` is such a good idea as the idea is to be able to rebuild it using the `go generate` command. It's like the `vendor` folder as far as I'm concerned.

But if you're having the problem, others will have it too, so it may be a good idea to commit it anyway. I'll think about it.

**Comment by prsolucoes** [comment]
I particularly dont have problems with commiting vendor folder. But this resource dont change by user and be part of your code. It is only a generated code from resources that never will change probably.

But it is my opinion, you decide.

Thanks.

**Comment by asticode** [comment]
Problem is `resources.go` also embeds the electron + astilectron releases, so I would need to generate it all over again everytime I bump their versions, which is a pain in the ass. I'll consider it.

**Comment by prsolucoes** [comment]
Ah, is embeds the entire zip to dont need download?

**Comment by prsolucoes** [comment]
I check the "resources.go" and it only have css/html. A small file of "8kb".

**Comment by prsolucoes** [comment]
And when i run, it try download the entire ZIP:

```
go run examples/9.bootstrap/main.go examples/9.bootstrap/resources.go -v
DEBU[0000] Restoring assets in /var/folders/jw/mq69589s7rj0r9bj9np1fk8m0000gn/T/go-build448139904/command-line-arguments/_obj/exe 
INFO[0000] You can run your custom provisioning here!   
DEBU[0000] Starting...                                  
DEBU[0000] Provisioning...                              
DEBU[0000] Provisioning Astilectron...                  
DEBU[0000] Removing directory /var/folders/jw/mq69589s7rj0r9bj9np1fk8m0000gn/T/go-build448139904/command-line-arguments/_obj/exe/vendor/astilectron 
DEBU[0000] Downloading https://github.com/asticode/astilectron/archive/v0.5.0.zip into /var/folders/jw/mq69589s7rj0r9bj9np1fk8m0000gn/T/go-build448139904/command-line-arguments/_obj/exe/vendor/astilectron-v0.5.0.zip 
2017-05-24 11:48:57.168 main[40724:7623217] *** Assertion failure in +[NSUndoManager _endTopLevelGroupings], /Library/Caches/com.apple.xbs/Sources/Foundation/Foundation-1349.81/Misc.subproj/NSUndoManager.m:363
2017-05-24 11:48:57.168 main[40724:7623217] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: '+[NSUndoManager(NSInternal) _endTopLevelGroupings] is only safe to invoke on the main thread.'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fffb01022cb __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x00007fffc4f0d48d objc_exception_throw + 48
	2   CoreFoundation                      0x00007fffb0107042 +[NSException raise:format:arguments:] + 98
	3   Foundation                          0x00007fffb1b4fc80 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 195
	4   Foundation                          0x00007fffb1ada323 +[NSUndoManager(NSPrivate) _endTopLevelGroupings] + 170
	5   AppKit                              0x00007fffadb674ed -[NSApplication run] + 1200
	6   main                                0x00000000043680fd uiMain + 29
	7   main                                0x00000000040566a0 runtime.asmcgocall + 112
)
libc++abi.dylib: terminating with uncaught exception of type NSException
DEBU[0000] Received signal abort trap                   
DEBU[0000] Stopping...                                  
signal: abort trap
```

We need a way to embed the ZIP to dont need download. And when i cross compile to other platforms each version need be it our ZIP. Cross compile for Linux need have his ZIP for Linux, For windows its ZIP for Windows, etc.

**Comment by asticode** [comment]
I was wrong, I inverted with the `5.single_binary_distribution` which embed the vendor directory. The `9.bootstrap` example don't so I've committed the `resources.go` file to ease running this example. I've also removed the loader which causes problems on MacOSX.

**Comment by prsolucoes** [comment]
But on example 5 we have another problem. How to generate the "vendor" folder?

**Comment by prsolucoes** [comment]
Sorry, i see here how to do it.

**Comment by prsolucoes** [comment]
How i can generate vendor.go in example 5 with embed files to other platform?

Example:

```
go generate examples/5.single_binary_distribution/main.go
cd examples/5.single_binary_distribution
env GOOS=windows GOARCH=amd64 go build
```

It generate a executable with 50mb +/-. But it has the ZIP for all platforms or only for MAC? Or Only for Windows?

**Comment by asticode** [comment]
As of now, it generates executables for your `runtime.GOOS` and `runtime.GOARCH`

**Comment by prsolucoes** [comment]
Im not asking about the executables man, im asking about the embed electron. The embed electron will be for mac/linux/windows or for all? Im asking about the "vendor.go" file.

**Comment by asticode** [comment]
The `vendor.go` file will contain what you have in the `../vendor` directory. The `../vendor` directory contains the executables for your `runtime.GOOS` and `runtime.GOARCH`. Therefore if you run with `GOOS=windows GOARCH=amd64`, the `../vendor` directory will contain **only** the executable for windows 64bits, and therefore the `vendor.go` will too.

As of now, it doesn't contain executables for all OS/ARCH couples. It contains executables for the `runtime.GOOS`/`runtime.GOARCH` you used when building the binary.

Does it make sense?

**Comment by prsolucoes** [comment]
I will try make all process here and test on my Virtual Machine with Windows.

**Comment by prsolucoes** [comment]
Lines executed:

```
cd
env GOOS=windows GOARCH=amd64 go generate main.go
env GOOS=windows GOARCH=amd64 go build
```

Files generated:

<img width="497" alt="screenshot 2017-05-24 13 20 38" src="https://cloud.githubusercontent.com/assets/395096/26413892/ce3872de-4083-11e7-8bd6-2545dbed6809.png">

See the error on Windows:
![virtualbox_windows 10_24_05_2017_13_18_45](https://cloud.githubusercontent.com/assets/395096/26413826/9ac867f6-4083-11e7-89c5-e95628851aeb.png)

I think the first problem is the version of astilectron, that is "astilectron-v0.3.0.zip". The other problems you need check. It is trying to extract to "c:/src" and not a temporary folder or a user folder.

**Comment by prsolucoes** [comment]
I have updated line to:

```
a.SetProvisioner(astilectron.NewDisembedderProvisioner(Asset, "../vendor/astilectron-v0.5.0.zip", "../vendor/electron-v1.6.5.zip"))
```

Generate the same things and the error changed to:

![virtualbox_windows 10_24_05_2017_13_27_01](https://cloud.githubusercontent.com/assets/395096/26414187/c8dc9c7e-4084-11e7-9db0-f18204956335.png)

**Comment by prsolucoes** [comment]
And after try execute the binary as administrator, i got this after 30seconds waiting:

![virtualbox_windows 10_24_05_2017_13_30_14](https://cloud.githubusercontent.com/assets/395096/26414321/31e2641a-4085-11e7-90b6-6c18ca77bfdf.png)

**Comment by prsolucoes** [comment]
Man, answering your questions: the theory make sense, but in practice no!

1 - You need use temp folder and go has a function to get it.
2 - You need embed correct ZIP as i said. Look the screenshot, you are sending MAC ZIP version, it is ignoring my GOOS and GOARCH variable.

![virtualbox_windows 10_24_05_2017_13_33_19](https://cloud.githubusercontent.com/assets/395096/26414526/d8aa3e44-4085-11e7-824c-2986eaad037e.png)

**Comment by prsolucoes** [comment]
Ah, and the third point, dont extract to a folder as go package. Use a folder that make sense for users and to be easy check to not redownloading, like:

```
${TMP}/asticode/vendor/electron-0.5.0/...
```

This is my suggestion.

**Comment by asticode** [comment]
1) indeed examples using `go-bindata` are not using the correct versions, I'll fix that and make it dynamic so that I don't have to change it anytime I bump either Astilectron or Electron's versions (because I will surely forget it)

2) `5.single_binary_distribution` was intended as an example which means that you should adapt it for your situation. Fact is I wanted people to be able to use `go run` with each example so I added the following line in each example `main.go`:

```go
var p = os.Getenv("GOPATH") + "/src/github.com/asticode/go-astilectron/examples"
```

That explains why it extracts to `c:/src` because `c:/` must be your **$GOPATH**. In that case, the example is working as intended (provide a way to run the example with `go run`).

3) You can choose the folder to extract to with the `BaseDirectoryPath` option.

4) I think we misunderstood each other, using `GOOS=windows GOARCH=amd64` in `GOOS=windows GOARCH=amd64 go generate main.go` is useless since it will embed what **is** in the vendor directory anyway. So if you just ran the project using `GOOS=darwin`, it will embed the MacOSX files. You have to remove what is in the `vendor` folder, **run** another example using `GOOS=windows GOARCH=amd64` (so that the vendor directory is populated with the correct files) and then run `go generate`. This will use `go-bindata` to simply embed what is in the `vendor` directory

Once again, this is an example on how to create a single binary distribution using `go-bindata`. **Astilectron** is not the one embedding data, `go-bindata` is. So you should take a look at this project if you want to go further.

5) Do you need a way to get all .zip files sorted by OS/ARCH so that you can generate single binary distribbutions for every OS/ARCH couples ?

**Comment by prsolucoes** [comment]
1 - I know that is an example, but if you do it in your main code, it will get automatically on real code or example. This will economize time to the developer. Auto embed correct versions for arch that you are compiling.

2 - Same as #1. If the code auto-get correct ZIP your examples can work on any system (mac, linux, windows). 

About path of extract, think on "user" not on "developer". I have installed a brand new windows on virtual box, and as a user, i double-click on program. It dont works, no message showed, and after execute from terminal (cmd), i have discovered the permission problem. Dont write to "C:", the default always need be "user home directory" or "temp directory", it will work for everyone.

3 - How i can generate the vendor path? I have deleted it and executed:
```
go generate examples/5.single_binary_distribution/main.go
```

But the folder are not recreated.

5 - As said on other topics, i think this process can be automatically and if not found, show error when execute "go build" or "go generate". It makes sense.

Others (remove following files, is irritating on mac):

- add ".DS_Store" to .gitignore
- add "Thumbs.db" to .gitignore
